File: boj_19637.cpp
#include <cstdio>
#include <cstring>

int count = 0;
char dict[100000][12] = {0,};
int score[100000] = {0,};

int binary(int target, int start, int end){

    if(start >= end){
        return end;
    }

    int mid = (start+end)/2;
    if(target <= score[mid]){
        if(mid > 0  && score[mid-1] < target){
            return mid;
        }
        else if(mid == 0){
            return mid;
        }
        else{
            return binary(target, start, mid);
        }
    }
    else{
        return binary(target, mid+1, end);
    }
}

int main(){
    int n,m; scanf("%d%d",&n,&m);

    for(int i=0;i<n;i++){
        int a;
        char t[12];
        scanf("%s",t);
        scanf("%d",&a);

        if( count >= 1 && 
            strcmp(t, dict[count]) == 0){
                continue;
        }
        strcpy(dict[count], t);
        score[count++] = a;
    }
    for(int i=0;i<m;i++){
        int a; scanf("%d",&a);
        printf("%s\n", dict[binary(a, 0, count-1)]);
    }

    return 0;
}

File: boj_19637.py

cnt_grade, cnt_user = [int(i) for i in input().split(" ")]
grade = []


def binary(start, end, target):
    
    if(start >= end):
        return grade[end][0]

    mid = int((start+end)/2)
    if target <= grade[mid][1]:
        if mid >= 1 and grade[mid-1][1] < target:
            return grade[mid][0]
        elif mid == 0:
            return grade[mid][0]
        else:
            return binary(start, mid, target)
    else:
        # grade[mid][1] < target
        return binary(mid+1, end, target)



for i in range(cnt_grade):
    g, standard = input().split(" ")
    if grade == [] or grade[-1][1] != int(standard):
        grade.append((g, int(standard)))


for _ in range(cnt_user):
    strength = int(input())
    print(binary(0, len(grade)-1, strength))


    

    
